packages/@lwc/ssr-runtime/src/clone-and-deep-freeze.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { freeze, entries, isObject, isArray, create, isNull } from '@lwc/shared';
+
+// Deep freeze and clone an object. Designed for cloning/freezing child props when passed from a parent to a child so
+// that they are immutable. This is one of the normal guarantees of both engine-dom and engine-server that we want to
+// emulate in ssr-runtime. The goal here is that a child cannot mutate the props of its parent and thus affect
+// the parent's rendering, which would lead to bidirectional reactivity and mischief.
+export function cloneAndDeepFreeze(obj: any): any {
+    if (isArray(obj)) {
+        const res = [];
+        for (let i = 0; i < obj.length; i++) {
+            res[i] = cloneAndDeepFreeze(obj[i]);
+        }
+        freeze(res);
+        return res;
+    } else if (!isNull(obj) && isObject(obj)) {
+        const res = create(null);
+        for (const [key, value] of entries(obj as any)) {
+            (res as any)[key] = cloneAndDeepFreeze(value);
+        }
+        freeze(res);
+        return res;
```

Comment:
Interesting strategy, going for a classic for loop for one but a for...of w/ destructuring for the other.

Reply:
Yeah I wasn't really going for consistency. I figure we will have to adjust this anyway because I imagine someone is relying on weird differences between `forEach`/`for` loops for hole-y arrays anyway. But I can make it a plain loop in both for now.